MoltenVK/MoltenVK/Commands/MVKMTLBufferAllocation.mm

```diff
@@ -39,16 +39,35 @@
     // of future allocation to beyond this allocation.
     NSUInteger offset = _nextOffset;
     _nextOffset += _allocationLength;
-    return new MVKMTLBufferAllocation(this, _mtlBuffers.back(), offset, _allocationLength);
+    return new MVKMTLBufferAllocation(this, _mtlBuffers.back().mtlBuffer, offset, _allocationLength, _mtlBuffers.size() - 1);
 }
 
 // Adds a new MTLBuffer to the buffer pool and resets the next offset to the start of it
 void MVKMTLBufferAllocationPool::addMTLBuffer() {
     MTLResourceOptions mbOpts = (_mtlStorageMode << MTLResourceStorageModeShift) | MTLResourceCPUCacheModeDefaultCache;
-    _mtlBuffers.push_back([_device->getMTLDevice() newBufferWithLength: _mtlBufferLength options: mbOpts]);
+    _mtlBuffers.push_back({ [_device->getMTLDevice() newBufferWithLength: _mtlBufferLength options: mbOpts], 0 });
     _nextOffset = 0;
 }
 
+MVKMTLBufferAllocation* MVKMTLBufferAllocationPool::acquireAllocation() {
+    if (_isThreadSafe) _lock.lock();
```

Comment:
Prefer to use `std::lock_guard`, because it guarantees unlocking when exiting scope, even in an exception or assertion. For example, something like:

     if (_isThreadSafe) {
        std::lock_guard<std::mutex> lock(_lock);
        return aquireAllocationUnsafely();
    } else {
        return aquireAllocationUnsafely();
    }

where `aquireAllocationUnsafely()` contains the remaining body of this function.

Reply:
Alright, there you go.